event_fout.c

```diff
@@ -6,11 +6,13 @@ typedef struct event_handler_fout_udata_s {
     char *cur;
     size_t buf_size;
     size_t rem;
+    int use_mutex;
```

Comment:
I'm not super familiar with the problem you're trying to solve, or how we run phpspy at etsy. But, just skimming I'll ask this: would using a mutex add latency (because of lock contention) in how we run phpspy? And would this potentially slow down web requests that we are tracing?

Reply:
I don't expect this to slow anything down. Tracing will get a little slower as threads will occasionally get blocked on the mutex. Requests will only slow down tinsofaras

`--rate-hz=499`

Good q. I think phpspy tracing is configured here:

https://github.com/etsy/go/blob/ce96914f1ed4c9cd1736a3beafa00717bbf52344/service/codesift_exporter/scraper_phpspy.yaml

Reply:
Traces can get interlaced, effectively corrupting the output, if they are bigger than 4k. This is because `write` is only atomic up to that value (at least on Linux). This is mentioned in the help output. The mutex is a way around this for the cost of an extra mutex.

The mutex is an extra resource for the kernel to manage, and tracing could slow down from the extra mutex-related syscalls and/or if threads are getting blocked. That said a single mutex shared by 15 or 20 threads should not make a dent on a production grade host.

Another way to solve this is to add an option to have each thread write to its own file instead of everyone writing to stdout. Then there's no contention. We could use that internally as well. It would require some refactoring in the services that wrap phpspy.

Reply:
i think i saw in the `phpspy` help output that the default number of threads was 16 (I think the default is used based on how phpspy is invoked in the codesift exporter [code](https://github.com/etsy/go/blob/e57a80fb2b5fdb815224be74fe0ce70a8b3d034a/service/codesift_exporter/main.go). So that's 16 out of 100 something threads on a host, which is not terrible. 

if we go the route of writing to a file, does that also slow down each request process (does writing to file take longer than writing to stdout)?

Reply:
I predict not unless we were writing with `O_DIRECT` which would bypass read/write caches on disks. We can measure it.